nima-databases-w2

Nimasaghi-dev opened the pull request.

dyaboykyl commented on Week2/exercise4.js:
```diff
@@ -0,0 +1,53 @@
+    try {
+        const queries = [
+            {
+                description: "All research papers and the id of authors",
```
The question in the instructions is not asking for the id of authors but the count of authors i.e. how many each paper has. That implies that one paper can have many authors, which should help you with comment above about the papers table. You will need to fix the papers table to remove the 1:many relationship and update this query, as this is currently assuming that every paper can only have 1 author

dyaboykyl commented on Week2/exercise4.js:
```diff
@@ -0,0 +1,53 @@
+            {
+                description: "Sum of the research papers published by all female authors",
+                query:
+                    "SELECT COUNT(research_paper.paper_id), author.gender FROM research_paper JOIN author ON research_paper.author_id = author.author_id WHERE author.gender = 'f';",
```
Same here and for others, these queries need to be updated to support a many-many relationship among papers and authors

dyaboykyl commented on Week2/exercise4.js:
```diff
@@ -0,0 +1,53 @@
+            {
+                description: "Average of the h-index of all authors per university",
+                query:
+                    "SELECT AVG(author.h_index), research_paper.university FROM author LEFT JOIN research_paper ON author.author_id = research_paper.author_id GROUP BY research_paper.university;",
```
This will only give you the average for each individual row which really doesn't mean anything. The question is asking for the average h_index **per university**, so you'll need a way to group all of the h_indexes for a university, before you take the average

Same goes for the below queries

dyaboykyl reviewed (changes requested):
The code looks great and is really clean, well done!

There's a couple of issues related to the relationships between authors/mentors and authors/papers, and the associated queries. There's also an issue with queries that ask for data _per university_. Please take a look a feel free to reach out on slack if you are unsure or stuck

dyaboykyl reviewed (changes requested):
Hey Nima, the code is getting closer to being fully correct, but there's a still some issues in how you define queries, as well as how you are inserting test data. Please take a look and don't hesitate to message me on slack if you are unsure of how to solve some of the problems!

Nimasaghi-dev commented:
> Hey Nima, the code is getting closer to being fully correct, but there's a still some issues in how you define queries, as well as how you are inserting test data. Please take a look and don't hesitate to message me on slack if you are unsure of how to solve some of the problems!

Hi @dyaboykyl , its getting more complicated each time i try to fix the problems, maybe it is better for me to start doing the assignment from scratch. what do you think ?

Nimasaghi-dev commented:
Hi @dyaboykyl ,I found the comments a bit overwhelming, which made it harder to follow along, so I decided to start the exercise from scratch to make it clearer and easier for me to understand.

dyaboykyl commented on Week2/db.sql:
```diff
@@ -0,0 +1,27 @@
+CREATE TABLE authors(
```
I think you don't need this file correct?

Nimasaghi-dev replied:
yes you're right, since i have it in the ex1.js

dyaboykyl commented on Week2/exercise4.js:
```diff
@@ -0,0 +1,43 @@
+        description: "Average H-index per university:",
+    },
+    {
+        query: `SELECT university, COUNT(research_papers.paper_id) AS paper_count FROM authors LEFT JOIN author_paper ON authors.author_id = author_paper.author_id LEFT JOIN research_papers ON author_paper.paper_id = research_papers.paper_id GROUP BY university;`,
```
1. What will happen if one paper has multiple authors from the same university? Will the results be correct?
2. You technically don't need need the second join on research_papers. You can get the paper_id from the author_paper table

Nimasaghi-dev replied:
If multiple authors from the same university worked on the same paper, the query might end up counting that paper more than once.

dyaboykyl commented on Week2/exercise2.js:
```diff
@@ -0,0 +1,42 @@
+    try{
+        await connection.query(createResearchPapersTable);
+        await connection.query(createAuthorPaperTable);
+        console.log("Research papers and author-paper tables created.");
```
Can you also add the fake data as requested in the instructions?

dyaboykyl reviewed (commented):
Hi Nima, well done! This is much cleaner that the first attempt and the relationships are formulated correctly. 

I just left a couple small comments

dyaboykyl commented on Week2/exercise1.js:
```diff
@@ -0,0 +1,102 @@
+    ('Author 15', 'University 5', '1995-05-05', 17, 'm');
+`;
+
+const insertResearchPapers = `
```
In order to fully verify your queries you need to add fake data that includes specifying with papers were written by which authors

Nimasaghi-dev commented on Week2/exercise1.js:
```diff
@@ -0,0 +1,102 @@
+
+connection.connect();
+
+const createResearchPapersTable = `
```
I accidentally removed it from both db.sql and ex1.js . :))

dyaboykyl reviewed (approved):
Thanks for addressing the comments I had, looks like you really cleaned things up!

Nimasaghi-dev commented:
> Thanks for addressing the comments I had, looks like you really cleaned things up!

Thanks for guiding me to fix the problems, it helped me to learn more.
